sources/api/bootstrap-commands/src/main.rs

```diff
@@ -0,0 +1,345 @@
+these commands, expanding the functionality of the existing Bottlerocket API.
+
+Currently, Bottlerocket provides a set of actions (such as refresh-updates and reboot) that can be
+accessed through the apiserver, typically by running the apiclient in the control container. However,
+this new feature aims to extend the accessibility of these actions by allowing Bottlerocket consumers
+to run them prior to the bootstrap container execution.
+
+#Example:
+Given a bootstrap command called `001-test-bootstrap-commands`  with the following configuration:
+
+[settings.bootstrap-commands.001-test-bootstrap-commands]
+commands = [[ "apiclient", "set", "motd=helloworld"]]
+essential = true
+mode = "always"
+
+This would update motd from "Welcome to Bottlerocket!" to "helloworld".
+*/
+
+#[macro_use]
+extern crate log;
+
+use itertools::Itertools;
+use serde::{Deserialize, Serialize};
+use simplelog::{Config as LogConfig, LevelFilter, SimpleLogger};
+use snafu::{ensure, OptionExt, ResultExt};
+use std::collections::HashMap;
+
+use std::env;
+use std::ffi::OsStr;
+use std::fs;
+use std::path::{Path, PathBuf};
+use std::process::{self, Command};
+use std::str::FromStr;
+
+use bottlerocket_modeled_types::{Identifier, BootstrapMode};
+use bottlerocket_settings_models::ApiclientCommand;
+
+const RUN_FILE_DIR: &str = "/run/bootstrap-commands";
+const DEFAULT_CONFIG_PATH: &str = "/etc/bootstrap-commands/bootstrap-commands.toml";
+const API_CLIENT_PATH: &str = "/usr/bin/apiclient";
+
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+struct BootstrapCommandConfig {
+    bootstrap_commands: Option<HashMap<Identifier, BootstrapCommand>>,
+}
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+struct BootstrapCommand {
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    commands: Option<Vec<ApiclientCommand>>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    mode: Option<BootstrapMode>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    essential: Option<bool>,
+}
+
+/// Stores user-supplied global arguments
+struct Args {
+    log_level: LevelFilter,
+    config_path: PathBuf,
+}
+
+impl Default for Args {
+    fn default() -> Self {
+        Self {
+            log_level: LevelFilter::Info,
+            config_path: PathBuf::from_str(DEFAULT_CONFIG_PATH).unwrap(),
+        }
+    }
+}
+
+/// Wrapper around process::Command that adds error checking.
+fn command<I, S>(bin_path: &str, args: I) -> Result<()>
+where
+    I: IntoIterator<Item = S>,
+    S: AsRef<OsStr>,
+{
+    let mut command = Command::new(bin_path);
+    command.args(args);
+    let output = command
+        .output()
+        .context(error::ExecutionFailureSnafu { command })?;
+
+    trace!("stdout: {}", String::from_utf8_lossy(&output.stdout));
+    trace!("stderr: {}", String::from_utf8_lossy(&output.stderr));
```

Comment:
If you do what I suggested above, you wouldn't need this.

Reply:
Updated.